services/requester/keystore.go

```diff
@@ -0,0 +1,111 @@
+package requester
+
+import (
+	"fmt"
+	"sync"
+
+	flowsdk "github.com/onflow/flow-go-sdk"
+
+	"github.com/onflow/flow-go-sdk/crypto"
+)
+
+var ErrNoKeysAvailable = fmt.Errorf("no keys available")
+
+type AccountKey struct {
+	flowsdk.AccountKey
+
+	mu      sync.Mutex
+	ks      *Keystore
+	Address flowsdk.Address
+	Signer  crypto.Signer
+	inuse   bool
+}
+
+// Done unlocks a key after use and puts it back into the pool.
+func (k *AccountKey) Done() {
+	k.markUnused()
+	k.ks.availableKeys <- k
+}
+
+// IncrementSequenceNumber is called when a key was successfully used to
+// sign a transaction as the proposer. It increments the sequence number.
+func (k *AccountKey) IncrementSequenceNumber() error {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	if !k.inuse {
+		return fmt.Errorf("key with index %d not locked", k.Index)
+	}
+
+	k.SequenceNumber++
+	return nil
+}
+
+func (k *AccountKey) SetProposerPayerAndSign(tx *flowsdk.Transaction) error {
+	return tx.
+		SetProposalKey(k.Address, k.Index, k.SequenceNumber).
+		SetPayer(k.Address).
+		SignEnvelope(k.Address, k.Index, k.Signer)
+}
+
+func (k *AccountKey) markUnused() {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	k.inuse = false
+}
+
+type Keystore struct {
```

Comment:
The benchmark Account keys/key store are actually poorly designed and cause issues :).

What you want from the `KeyStore` is just a public `Take` (or Get) method. The key store should also have a reference to a block Publisher `*models.Publisher[*models.Block]`.

The usage code should look something like this:

```go
func someFuncCreatingAndSendingATX( keyStore KeyStore) {
   accountKey, err := keyStore.Take()
   // handle err
   defer accountKey.ReturnIfNotReserved()

   referenceBlock := getReferenceBlock()
   
   // ...
   tx, err := createAndSignTX(accountKey, referenceBlock.Height)
   if err != nil {
     // defer will handle cases where we haven't actually used the key
     return
   }

   txID, err := sendTX(tx)
   if err != nil {
     return
   }
   accountKey.ReserveFor(txID,referenceBlock.Height)
 
}
```

the `ReserveFor` would do something like:

```go
func (k *AccountKey) ReserveFor(txID, referenceHeight) {
  k.keystore.blockPublisher.
  // create new subscritpion that
  // if txID is found. increments account key sequence number
  // if referenceHeight is more than 1000 blocks past doesnt increment the sequence number
  // then unsubsribes and returns the key to the keystore
}

```

`ReturnIfNotReserved` would just check if the key is reserved, if not it would just return it to the keystore.

Reply:
The `referenceBlock.Height` refers to Flow block heights, while `*models.Publisher[*models.Block]` deals with EVM block heights. So I am not very fond of using this approach, given that we might experience failures from the system chunk tx. Also, we only have the `txID` for the submitted Flow transaction. We can't know if this will end up creating an EVM transaction.

Reply:
I have updated the implementation in this commit: https://github.com/onflow/flow-evm-gateway/pull/693/commits/2a7482e3de5ded6bb01faad50d2e41fc788da628 . Let me know what you think :pray:

Reply:
My main concern with this is that if the sequence number de-syncs, its going to stay de-synced, as there is no correction mechanism. One fix to this is a more robust way to keep them synced, and the other is to detect de-syncs and fix them.

The sequence number could desync for a few reasons:
- transaction failed in a way where the sequence number does not get incremented, and the gateway still incremented it
- the gateway thinks the transaction was lost (and doesn't increment the sequence number), but it is just late (or somehow missed)
- user manually used one of the keys for a transaction
- ... (maybe something I'm forgetting)

Using the time for detection is not guaranteed to work. As the chain might be slow or down and the transaction will succeed a bit later, reusing that key in the meantime will produce a failing transaction and a de-sync of the sequence number. With the height based approach we know that if the transaction hasn't gone through in X amount of blocks, its not going to go through, because the reference block is too old. Good point with the `*models.Publisher[*models.Block]` maybe we would need a publisher for flow blocks as well then.

as for the auto-correction I would only increment the sequence number after the results of the transactions are known. If the transaction resulted in a sequence number error I would re-fetch the sequence number for that key instead of incrementing it. If the transaction expired I just would not increment the sequence number.

Another (easier/shorter/temporary) solution for this problem would be to crash the whole gateway if a transaction results in a sequence number mismatch.

Reply:
I've actually reworked the whole in-memory approach for `SequenceNumber`. See: https://github.com/onflow/flow-evm-gateway/pull/693/commits/74cb17da51f3e3d0e67f761135c37a6c9e230bf3#diff-c560dc90897114da6ccaa724718c1055144a1c195676fb1e047c0200899aecb9R35-R53 .

Reply:
I will also try to find a way for auto-expiring based on the reference block height :+1:

Reply:
@janezpodhostnik I added an approach for expiring account keys based on reference block height here: https://github.com/onflow/flow-evm-gateway/pull/693/commits/374f70c9e6393ad95174fc67f0d85d6b4e3ed9aa . I guess it's not something that we have to do for every single Flow block. But let me know what you think.

Reply:
nice, this will work. we can make it more optimal later if need be.